### PR TITLE
♻️  Types: opt for null shorthand

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Opts to use null shorthand when not too disruptive, based [on earlier feedback](https://github.com/ampproject/amphtml/pull/34105#discussion_r626667589)

### CODE REVIEW

File: build-system/babel-plugins/babel-plugin-transform-fix-leading-comments/test/fixtures/transform/basic/output.js [hunk 1]
```diff
@@ -15,7 +15,7 @@
  */
 
 /**
- * @protected {!Object<string, !ChildDef>|null}
+ * @protected {?Object<string, !ChildDef>}
```
Comment (rcebulko) [inline build-system/babel-plugins/babel-plugin-transform-fix-leading-comments/test/fixtures/transform/basic/output.js:18]: `Object` is implicitly nullable
```suggestion
 /** @protected {Object<string, !ChildDef>} */
```

File: build-system/common/utils.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ async function buildRuntime(opt_compiled = false) {
 /**
  * Extracts and validates the config for the given experiment.
  * @param {string} experiment
- * @return {Object|null}
+ * @return {?Object}
```
Comment (rcebulko) [inline build-system/common/utils.js:49]:
```suggestion
 * @return {Object}
```

File: build-system/externs/amp.extern.js [hunk 3]
```diff
@@ -624,7 +624,7 @@ AmpElement.prototype.collapse = function () {};
 let Signals = class {};
 /**
  * @param {string} unusedName
- * @return {number|!Error|null}
+ * @return {?number|?Error}
```
Comment (rcebulko) [inline build-system/externs/amp.extern.js:627]: primitives can never be nullable
```suggestion
 * @return {number|?Error}
```

File: build-system/server/app.js [hunk 4]
```diff
@@ -1560,7 +1560,7 @@ function generateInfo(filePath) {
 
 /**
  * @param {string} encryptedDocumentKey
- * @return {string|null}
+ * @return {?string}
```
Comment (rcebulko) [inline build-system/server/app.js:1563]: primitives themselves can't be nullable; need explicit syntax (how it is now)
```suggestion
 * @return {string|null}
```

File: build-system/tasks/markdown-toc/index.js [hunk 5]
```diff
@@ -84,7 +84,7 @@ function isolateCommentJson(maybeComment) {
 
 /**
  * @param {string} content
- * @return {Promise<string|null>}
+ * @return {Promise<?string>}
```
Comment (rcebulko) [inline build-system/tasks/markdown-toc/index.js:87]:
```suggestion
 * @return {Promise<string|null>}
```

File: build-system/tasks/performance/stats.js [hunk 6]
```diff
@@ -86,7 +86,7 @@ const q3 = (array) => {
  *
  * @param {number} a
  * @param {number} b
- * @return {number|null} percentage change or null
+ * @return {?number} percentage change or null
```
Comment (rcebulko) [inline build-system/tasks/performance/stats.js:89]:
```suggestion
 * @return {number|null} percentage change or null
```

**Review by rcebulko — CHANGES_REQUESTED** [review]
For any primitives (ie. `boolean`, `string`, `number`, `undefined`, and `function` iirc) nullable ? syntax isn't supported, `|null` required (this should probably be used for `{T|null}` now that I think about it)

File: build-system/babel-plugins/babel-plugin-transform-fix-leading-comments/test/fixtures/transform/basic/output.js (hunk 1 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-fix-leading-comments/test/fixtures/transform/basic/output.js:18]: It is. But if you [read the docs](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#the-javascript-type-language) under "Nullable Type", they recommend being explicit and including the `?` anyway.

**Comment by samouri** [comment]
> For any primitives (ie. `boolean`, `string`, `number`, `undefined`, and `function` iirc) nullable ? syntax isn't supported, `|null` required (this should probably be used for `{T|null}` now that I think about it)

@rcebulko: check [the docs](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#the-javascript-type-language) under "Nullable types". Unlike in other languages, primitive types can be nullable.

File: ads/google/a4a/utils.js [hunk 7]
```diff
@@ -523,7 +523,7 @@ function makeCorrelator(pageViewId, opt_clientId) {
 /**
  * Collect additional dimensions for the brdim parameter.
  * @param {!Window} win The window for which we read the browser dimensions.
- * @param {{width: number, height: number}|null} viewportSize
+ * @param {?{width: number, height: number}} viewportSize
```
[the remark this answers is not in the document]
Comment (samouri) [reply ads/google/a4a/utils.js:526]: I'd rather you make a GH Issue. The type for `{width: number, height: number}` is super common in the codebase, and leaving it here in this file is somewhat random.

[the remark this answers is not in the document]
Comment (samouri) [reply ads/google/a4a/utils.js:526]: Also note: we already have multiple extern definitions for the type as well.
See `DimensionsDef` and `LayoutSizeDef`

**Comment by rcebulko** [comment]
> > For any primitives (ie. `boolean`, `string`, `number`, `undefined`, and `function` iirc) nullable ? syntax isn't supported, `|null` required (this should probably be used for `{T|null}` now that I think about it)
> 
> @rcebulko: check [the docs](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#the-javascript-type-language) under "Nullable types". Unlike in other languages, primitive types can be nullable.
> 
> > Object is implicitly nullable
> 
> It is. But the docs recommend being explicit and including the `?` anyway.

I believed the exact opposite! Thanks for pointing me to that page

**Review by jridgewell — APPROVED** [review]

**Review by rcebulko — APPROVED** [review]

**Pull request merged**